### ISSUE TITLE
Compiling NativeOps does not work anymore with GCC-5

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#890]
When compiling `NativeLstm2`I get the following error:
```
/work/tools/asr/python/3.8.0_tf_2.3.4-haswell+cuda10.1+mkl/lib/python3.8/site-packages/tensorflow/include/absl/time/internal/cctz/include/cctz/civil_time_detail.h:587:20: error: call to non-constexpr function ‘absl::lts_2020_02_25::time_internal::cctz::detail::civil_time<absl::lts_2020_02_25::time_internal::cctz::detail::day_tag> absl::lts_2020_02_25::time_internal::cctz::detail::operator-(absl::lts_2020_02_25::time_internal::cctz::detail::civil_time<absl::lts_2020_02_25::time_internal::cctz::detail::day_tag>, absl::lts_2020_02_25::time_internal::cctz::diff_t)’ 
           return cd - (j - i); 
                    ^ 
```

This is a known bug because GCC-5 is not correctly (at least not in the sense of how the `absl` library wants it) handling the c++14 mode. In c++11 constexpr are disabled, but we added this specific hack to disable it for c++14 as well:

```
diff --git a/absl/time/internal/cctz/include/cctz/civil_time_detail.h b/absl/time/internal/cctz/include/cctz/civil_time_detail.h
--- a/absl/time/internal/cctz/include/cctz/civil_time_detail.h
+++ b/absl/time/internal/cctz/include/cctz/civil_time_detail.h
@@ -23,7 +23,7 @@
 #include "absl/base/config.h"
 
 // Disable constexpr support unless we are in C++14 mode.
-#if __cpp_constexpr >= 201304 || (defined(_MSC_VER) && _MSC_VER >= 1910)
+#if (!defined(NO_CONSTEXPR_FOR_YOU) && __cpp_constexpr >= 201304) || (defined(_MSC_VER) && _MSC_VER >= 1910)
 #define CONSTEXPR_D constexpr  // data
 #define CONSTEXPR_F constexpr  // function
 #define CONSTEXPR_M constexpr  // member
```

So what we (at i6) could do now is:
 - add this `NO_CONSTEXPR_FOR_YOU` flag to the compile opts (very specific, would only work for custom compiled TF versions but not for default pip installed ones)
 - default to c++11 also for GCC-5 (see #875 )
 - Not updating RETURNN anymore until we switch to Ubuntu 20 (which brings the correct GCC versions that Tensorflow is tested against)
 
This is quite the serious issue, as since the 6ee499ac7b3222db058f8c1e6cad19ce7875a056 (#826) update from 6th December no TF2 based setup that uses native ops will work (at i6).

**Comment by albertz** [comment]
> When compiling `NativeLstm2`I get the following error:

Is this really about NativeLSTM2 or any native op compilation in general?

> ```
> /work/tools/asr/python/3.8.0_tf_2.3.4-haswell+cuda10.1+mkl/lib/python3.8/site-packages/tensorflow/include/absl/time/internal/cctz/include/cctz/civil_time_detail.h:587:20: error: call to non-constexpr function ‘absl::lts_2020_02_25::time_internal::cctz::detail::civil_time<absl::lts_2020_02_25::time_internal::cctz::detail::day_tag> absl::lts_2020_02_25::time_internal::cctz::detail::operator-(absl::lts_2020_02_25::time_internal::cctz::detail::civil_time<absl::lts_2020_02_25::time_internal::cctz::detail::day_tag>, absl::lts_2020_02_25::time_internal::cctz::diff_t)’ 
>            return cd - (j - i); 
>                     ^ 
> ```

This is a very old TF version. Did you try some more recent one? Maybe it works there?

> This is a known bug because GCC-5 is not correctly (at least not in the sense of how the `absl` library wants it) handling the c++14 mode. In c++11 constexpr are disabled, but we

Who is *we* here?

> added this specific hack to disable it for c++14 as well:
> 
> ```
> diff --git a/absl/time/internal/cctz/include/cctz/civil_time_detail.h b/absl/time/internal/cctz/include/cctz/civil_time_detail.h
> --- a/absl/time/internal/cctz/include/cctz/civil_time_detail.h
> +++ b/absl/time/internal/cctz/include/cctz/civil_time_detail.h
> @@ -23,7 +23,7 @@
>  #include "absl/base/config.h"
>  
>  // Disable constexpr support unless we are in C++14 mode.
> -#if __cpp_constexpr >= 201304 || (defined(_MSC_VER) && _MSC_VER >= 1910)
> +#if (!defined(NO_CONSTEXPR_FOR_YOU) && __cpp_constexpr >= 201304) || (defined(_MSC_VER) && _MSC_VER >= 1910)
>  #define CONSTEXPR_D constexpr  // data
>  #define CONSTEXPR_F constexpr  // function
>  #define CONSTEXPR_M constexpr  // member
> ```

You should not add such a hack. You should also not edit the TF sources.

> So what we (at i6) could do now is:
> 
> * add this `NO_CONSTEXPR_FOR_YOU` flag to the compile opts (very specific, would only work for custom compiled TF versions but not for default pip installed ones)

Why so complicated? Why not just sth like `-Dconstexpr=`? I think we also did that in the past, as far as I remember. Although that is also bad.

> * default to c++11 also for GCC-5 (see [Use -std=c++11 for gcc v4 and lower #875](https://github.com/rwth-i6/returnn/pull/875) )

So, is there any problem with this approach?

> * Not updating RETURNN anymore until we switch to Ubuntu 20 (which brings the correct GCC versions that Tensorflow is tested against)

I don't really understand why this is an option, or why you even mention that here?

**Comment by JackTemaki** [comment]
> > When compiling `NativeLstm2`I get the following error:
> 
> Is this really about NativeLSTM2 or any native op compilation in general?

I don't know. It could be a NativeLSTM2 specific import/include but it could be for all.

> This is a very old TF version. Did you try some more recent one? Maybe it works there?

I do not have never versions (And I do not plan to spend time on this before the OS change), but according to https://github.com/tensorflow/tensorflow/issues/41926 this problem still does exist at least in TF2.5.

> > This is a known bug because GCC-5 is not correctly (at least not in the sense of how the `absl` library wants it) handling the c++14 mode. In c++11 constexpr are disabled, but we
> 
> Who is _we_ here?

Everybody that uses TF2 with RASR and GCC-5, I think Eugen introduced this some time ago.

> > added this specific hack to disable it for c++14 as well:
> > ```
> > diff --git a/absl/time/internal/cctz/include/cctz/civil_time_detail.h b/absl/time/internal/cctz/include/cctz/civil_time_detail.h
> > --- a/absl/time/internal/cctz/include/cctz/civil_time_detail.h
> > +++ b/absl/time/internal/cctz/include/cctz/civil_time_detail.h
> > @@ -23,7 +23,7 @@
> >  #include "absl/base/config.h"
> >  
> >  // Disable constexpr support unless we are in C++14 mode.
> > -#if __cpp_constexpr >= 201304 || (defined(_MSC_VER) && _MSC_VER >= 1910)
> > +#if (!defined(NO_CONSTEXPR_FOR_YOU) && __cpp_constexpr >= 201304) || (defined(_MSC_VER) && _MSC_VER >= 1910)
> >  #define CONSTEXPR_D constexpr  // data
> >  #define CONSTEXPR_F constexpr  // function
> >  #define CONSTEXPR_M constexpr  // member
> > ```
> 
> You should not add such a hack. You should also not edit the TF sources.

Should not, but nobody found a different solution so far.

> > So what we (at i6) could do now is:
> > 
> > * add this `NO_CONSTEXPR_FOR_YOU` flag to the compile opts (very specific, would only work for custom compiled TF versions but not for default pip installed ones)
> 
> Why so complicated? Why not just sth like `-Dconstexpr=`? I think we also did that in the past, as far as I remember. ([tensorflow/tensorflow#22766](https://github.com/tensorflow/tensorflow/issues/22766)) Although that is also bad.

That does not work.

> > * default to c++11 also for GCC-5 (see [Use -std=c++11 for gcc v4 and lower #875](https://github.com/rwth-i6/returnn/pull/875) )
> 
> So, is there any problem with this approach?

Probably not.

> > * Not updating RETURNN anymore until we switch to Ubuntu 20 (which brings the correct GCC versions that Tensorflow is tested against)
> 
> I don't really understand why this is an option, or why you even mention that here?

Because I am spending a lot of time on fixing technical issues for an environment that we might completely replace this year, so I am not sure if this is a reasonable thing to do (trying to fix RETURNN/Tensorflow with respect to officially unsupported build environments).

**Comment by albertz** [comment]
> > You should not add such a hack. You should also not edit the TF sources.
> 
> Should not, but nobody found a different solution so far.

But surely we knew a couple of better solutions already before, and also you and I suggested some above.

> > Why so complicated? Why not just sth like `-Dconstexpr=`? I think we also did that in the past, as far as I remember. ([tensorflow/tensorflow#22766](https://github.com/tensorflow/tensorflow/issues/22766)) Although that is also bad.
> 
> That does not work.

Why?

Well, but anyway, you already found an even simpler solution (`-std=c++11`), so I don't really understand the problem.

> > > default to c++11 also for GCC-5 (see [Use -std=c++11 for gcc v4 and lower #875](https://github.com/rwth-i6/returnn/pull/875) )
> > 
> > So, is there any problem with this approach?
> 
> Probably not.

So what is the problem then?

> > > Not updating RETURNN anymore until we switch to Ubuntu 20 (which brings the correct GCC versions that Tensorflow is tested against)
> > 
> > I don't really understand why this is an option, or why you even mention that here?
> 
> Because I am spending a lot of time on fixing technical issues for an environment that we might completely replace this year, so I am not sure if this is a reasonable thing to do (trying to fix RETURNN/Tensorflow with respect to officially unsupported build environments).

But surely whatever you do must work with RETURNN master. Also, I don't really see any problems in just fixing the issue. The issue seems extremely trivial to fix?

**Comment by albertz** [comment]
Should be fixed by #896.
